src/lib.rs

```diff
@@ -1,202 +1,284 @@
-                }))
-                .await
-                .unwrap();
-                tokio::time::sleep(Duration::from_secs(1)).await;
-                tx.send(Events::DataReceived(DataReceived {
-                    connection_id: 42,
-                    data: vec![0, 1, 2],
-                }))
-                .await
-                .unwrap();
-                tokio::time::sleep(Duration::from_secs(1)).await;
-                tx.send(Events::ConnectionClosed(ConnectionClosed { connection_id: 42 }))
-                    .await
-                    .unwrap();
-            }
-        });
+    pub async fn new(
+        host: String,
+        port: u16,
+        private_key: String,
+        peer_public_keys: Vec<String>,
+        handle_connection: PyObject,
+        receive_datagram: PyObject,
+    ) -> Result<WireguardServer> {
+        let private_key: Arc<X25519SecretKey> = Arc::new(private_key.parse().map_err(|error: &str| anyhow!(error))?);
+        let peers = peer_public_keys
+            .into_iter()
+            .map(|peer| {
+                let key = Arc::new(X25519PublicKey::from_str(&peer).map_err(|error: &str| anyhow!(error))?);
+                Ok((key, None))
+            })
+            .collect::<Result<Vec<(Arc<X25519PublicKey>, Option<[u8; 32]>)>>>()?;
+
+        let (py_loop, run_coroutine_threadsafe) = Python::with_gil(|py| -> PyResult<(PyObject, PyObject)> {
+            Ok((
+                pyo3_asyncio::tokio::get_current_loop(py)?.into(),
+                py.import("asyncio")?.getattr("run_coroutine_threadsafe")?.into(),
+            ))
+        })?;
+
+        let (wg_to_smol_tx, wg_to_smol_rx) = channel(16);
+        let (smol_to_wg_tx, smol_to_wg_rx) = channel(16);
+
+        let (smol_to_py_tx, mut smol_to_py_rx) = channel(64);  // only used to notify of incoming connections and datagrams
+        let (py_to_smol_tx, py_to_smol_rx) = unbounded_channel();  // used to send data and to ask for packets. We need this to be unbounded as write() is not async.
+
+        let mut wg_server =
+            wireguard::WireguardServer::new((host, port), private_key, peers, wg_to_smol_tx, smol_to_wg_rx).await?;
+        let local_addr = wg_server.local_addr()?;
 
+        let mut tcp_server = tcp::TcpServer::new(smol_to_wg_tx, wg_to_smol_rx, smol_to_py_tx, py_to_smol_rx)?;
+
+        let wireguard_task = tokio::spawn(async move { wg_server.run().await });
+        let tcp_task = tokio::spawn(async move { tcp_server.run().await });
+
+        let event_tx = py_to_smol_tx.clone();
         // this task feeds events into the Python callback.
-        let call_python_callback = tokio::spawn(async move {
-            while let Some(event) = rx.recv().await {
-                Python::with_gil(|py| {
-                    if let Err(err) = on_event.call1(py, (event,)) {
-                        err.print(py);
-                    }
-                });
+        let python_notify_task = tokio::spawn(async move {
```

Comment:
Architecturally we only have these three main tasks. We do not spawn any subtasks for individual clients or so.

`wireguard_task`: udp -> network events
`tcp_task`: network events -> transport events
`python_notify_task`: transport events -> notify python land